docs/extensions/readme.md

```diff
@@ -1173,6 +1174,112 @@ When set, specifies the header parameter to be used instead of `x-ms-client-requ
 }
 ```
 
+## x-ms-allowed-arm-id-types
+
+Can only be set on `"type": "string"` fields with `"format": "arm-id"`.
+When set, specifies the set of resource types which can be referenced by this `arm-id`. If this extension isn't provided for a particular `arm-id`, the field can refer to any valid ARM ID.
+
+**Parent element**: [Parameter Object](https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#parameterObject), [Schema Object](https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#schemaObject), or [Items Object](https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#itemsObject)
+
+**Schema**:
+
+| Field Name    | Type       | Description                                      |
+| ------------- | ---------- | ------------------------------------------------ |
+| allowedTypes  | `[string]` | **Required** An array of allowed ARM resource ID kinds. Each element represents a particular type of ARM resource which can be referred to by this `arm-id`. See [Allowed formats](#allowed-arm-id-formats) for a detailed descrition of the allowed formats. |
+
+### Allowed ARM ID formats
+Note: When reading the formats below, `{}` is left as-is in the `allowedTypes` entry. Named fields such as `{provider}` must be given concrete values in the `allowedTypes` entry.
+
+| Resource kind                   | Format                                                                      | Example `allowedTypes` entry                                                       |
+| ------------------------------- | --------------------------------------------------------------------------- | ---------------------------------------------------------------------------------- |
+| Resource in a tenant            | `/providers/{provider}/{resourceType}`                                      | `/providers/Microsoft.Capacity/reservationOrders`                                  |
+| Resource in a subscription      | `/subscriptions/{}/providers/{provider}/{resourceType}`                     | `/subscriptions/{}/providers/Microsoft.Resources/resourceGroups`                   |
```

Comment:
I originally wanted to use the same format that ARM specifies in ARM templates (`Microsoft.Resources/resourceGroups` or `Microsoft.Compute/virtualMachines`, etc), but that format doesn't actually convey if the resource roots up to subscription, resourceGroup, or something else, which is required to validate resource ID validity (or generate an ID that mathces) so I think we need a more verbose format such as this one.

Reply:
One id property MAY want to point to specified resource type but don't want to specify scope. For example, some property may want to point to any authorization resource. How do you think about this?

Reply:
@PhoenixHe-msft - what does an authorization resource URL look like?

Reply:
I assume that the general question is what referring to an [extension resource](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/extension-resource-types) looks like... In an ARM template, this would be done by providing the [scope](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/scope-extension-resources) for the resource.

I'm not sure how deep down the rabbit hole we'd like to run with specifying that an ARM ID can only refer to extension resource type A but only if it is scoped to resource type B. I believe that could be a recursive definition that would make the hole very deep indeed. But I could see a `scope: *` or similar added to indicate that this is an extension provider.

Reply:
Exactly. Suggest to referencing the design of ARM template's scope.

Reply:
From my read of the scope documentation (Thanks @johanste) it looks like scope is inferred in a large number of cases based on the other fields specified.

This seems to be confirmed by the definition of the `scope` field in the JSON Schema:
```
            "scope": {
              "type": "string",
              "description": "Scope for the resource or deployment. Today, this works for two cases: 1) setting the scope for extension resources 2) deploying resources to the tenant scope in non-tenant scope deployments"
            },
```

So that means that the valid values of `scope` are:
1. `/`
2. Something like `Microsoft.Storage/storageAccounts/myStorageAccountName` in the case of extension resources.

This list of values doesn't provide us the level of specification that we need, because it's not actually saying if it's a RG or Subscription (or region) scoped deployment. They instead infer that detail from the other template fields (which we don't have here).

So we would need to augment for our usage of scope (if we're going to use it that is).

Are you suggesting that we change this to something like the below?

```json5
"MyExampleType": {
  "properties": {
    "vnetId": {
      "type": "string",
      "format": "arm-id",
      "x-ms-arm-id-details": {
        "allowedTypes": [
          {
            "scope": "/subscriptions/resourceGroups",
            "resource": "Microsoft.Network/virtualNetworks"
          }
        ]
      }
    }
  }
}
```

We'd have to discuss what values we would allow for `scope` because as mentioned above the set that ARM supports isn't sufficient for our needs. If people do prefer this approach though we could have:

1. `/` - Tenant
2. `/subscriptions` - Subscription
3. `/subscriptions/resourceGroups` - ResourceGroup
4. `*` - For extension resources

Reply:
Or are you saying a single field string and just make it "look like a scope" on the left hand side:
`/subscriptions/providers/Microsoft.Resources/resourceGroups` - everything to the left hand side of `providers` is the scope.

Reply:
Sorry for the delay -- I've pushed an update to this PR that moved towards the recommended `scope` style.

I went with the shorter "all in a single string" format over the longer `"scope"` + `"resource"` properties above as it felt cleaner and parsing the single string to determine the two pieces (scope and resource) should be easy.

Let me know what you think @johanste and @PhoenixHe-msft

Reply:
@majastrz for comments, who may have valuable comments for the way of specifying resource types and scopes, as he is working on another proposal `x-ms-resource-scope`.

Reply:
@meni-msft great catch. I think the key differences in @majastrz proposal and this one are:

1. It supports one more scope (management groups) at `/providers/Microsoft.Management/managementGroups`. I believe I should also support this.
2. It proposes two possible scope formats that are somewhat different than the one we've landed on here. We should unify on a format.
3. It supports `AnyResource` and `AnyResourceExtension` in addition to `*`. We should decide if we want/need `AnyResource`/`AnyResourceExtension` here as well. @majastrz pointed out that we could also add these later if we just started with `*` so this isn't a hard blocker I don't think.

From this list I think the key question is on item 2 above, how the `scope` section should be specified.

Reply:
@matthchr For item 2, to unify the scope format with @majastrz 's proposal, shall we change the extension like below?
```
"MyExampleType": {
  "properties": {
    "id": {
      "type": "string",
      "format": "arm-id",
      "x-ms-arm-id-details": {
        "allowedResources": [
          {
            "scopes": ["Tenant", "Subscription", "ResourceGroup"],
            "type": "Microsoft.Network/virtualNetworks"
          },
          {
            "scopes": ["Tenant", "Subscription", "ResourceGroup"],
            "type": "Microsoft.Compute/disks"
          },
          {
            "type": "Microsoft.Compute/virtualMachines"
          }
        ]
      }
    }
  }
}
```
The possible values of `scopes` are [ `Tenant`, `Subscription`, `ResourceGroup`, `ManagementGroup`, `*` ], and make `scopes` optional and default to `["ResourceGroup"]`.

So the changes include:
* `allowedTypes` -> `allowedResources`
* add `scopes`
* `resource` -> `type`

Since this is Azure-specific swagger extension, it should be fair to use these and `resourceType`.

Reply:
- `allowedTypes` -> `allowedResources`: I'm happy to make this change.
- add `scopes`: Agreed
- `resource` -> `type`: The other document calls it `paths` (or `path` for singular here), and it uses the full URL like `/subscriptions/{subscriptionId}/resourceGroups/{rg}/providers/Microsoft.Compute/virtualMachines/{vm}` whereas your example us just using `Microsoft.Compute/virtualMachines`. Just wanted to highlight that difference to the proposal that @majastrz made. I do think what you're suggesting here makes a lot of sense though as it doesn't make sense to put the full path here if it can possibly apply to many scopes (since the full path would be different for each scope).

I'll go ahead and make these changes because they seem good to me.

Reply:
Oh one more thing, I think `type` probably should support `*` too. That way you can do something like:
```
// This can be any ARM ID  
"id": {
  "type": "string",
  "format": "arm-id",
}
```

```
// This can be any ARM ID as long as it's RG scoped
"id": {
  "type": "string",
  "format": "arm-id",
  "x-ms-arm-id-details": {
    "allowedResources": [
      {
        "scopes": ["ResourceGroup"],
        "type": "*"
      },
    ]
}
```

I don't actually know if we have any resources that need this functionality today but I don't know that we don't so we might as well support it?

Reply:
One other thought... I think we should drop `*` and replace it with `Extension`.

`*`'s meaning is ambiguous right now. Is `"scope": "*"` == `"scope": ["Tenant", "Subscription", "ResourceGroup", "ManagementGroup"]`? Not really, because `*` also supports extension resources. It's confusing if `"scope": "*"` just means "this resource is an extension resource" though - if we want it to mean that we should call it `"scope": "Extension"` or `"scope": "AnyResource"` instead.

`*` meaning "All scopes" doesn't seem that useful because `scopes` is already a collection (as proposed by @leni-msft), and it seems unlikely to me that if Azure comes up with a new scope that all the old resources that said `"scope": "*"` with actually jump to immediately supporting references at the new scope. It seems better to just be explicit and make resources that refer to something like `Microsoft.Resources/deployments` say:
```
"deploymentId": {
  "type": "string",
  "format": "arm-id",
  "x-ms-arm-id-details": {
    "allowedResources": [
      {
        "scopes": ["Tenant", "Subscription", "ResourceGroup", "ManagementGroup"],
        "type": "Microsoft.Resources/deployments"
      },
    ]
}
```

Thoughts?

Reply:
Thanks @matthchr for the quick updates.

> * `resource` -> `type`: The other document calls it `paths` (or `path` for singular here) ... 

No problem, the change has also been agreed by @majastrz in the doc comment.

> I think type probably should support `*` too

Agree. One further step, the type should support `Microsoft.Compute/*` - to limit the resource providers but allow any resource types under it.

> I think we should drop `*` and replace it with `Extension`

Agree to add `Extension`. But I suggest to keep `*`(any valid scope) just for easier writing - it's unlikely that Azure will have a new scope.

Reply:
> Agree. One further step, the type should support Microsoft.Compute/* - to limit the resource providers but allow any resource types under it.

Done.

> Agree to add Extension as a valid scope value. But I suggest to keep *(any valid scope) just for easier writing - it's unlikely that Azure will have a new scope.

Ok, as long as it's not likely for Azure to be adding new scopes I'm fine with `*`. Have added it back and documented it.